### PR TITLE
support depot mapping and directory mapping for document id generation

### PR DESCRIPTION
**Pull request by OsmondJiang**

### CODE REVIEW

**Comment by OsmondJiang** [comment]
#2821

File: docs/specs/metadata.yml [hunk 1]
```diff
@@ -99,3 +99,31 @@ outputs:
...
     ["warning","redirected-id-conflict","Multiple documents redirected to '/sccm/mdm/understand/hybrid-mobile-device-management' with document id: 'sccm/mdm/index.md', 'sccm/mdm/index2.md'","docfx.yml"]
+---
+# generate document id with depot mapping and directory mapping
+# id = hash({mapped_depot_name}|{mapped_file_in_docset})
+# version_independent_id = hash({mapped_depot_name}|{file_path})
+inputs:
+  docfx.yml: |
+    documentId:
+      siteBasePath: azure
+      depotMappings:
```
Comment (yufeih) [inline docs/specs/metadata.yml]: can we keep the original `_op_documentIdPathDepotMapping` syntax? `file -> path + depot`

File: src/docfx/config/DocumentIdConfig.cs [hunk 2]
```diff
@@ -1,10 +1,17 @@
...
+using System;
+using System.Collections.Generic;
+using System.IO;
+
 namespace Microsoft.Docs.Build
 {
     internal class DocumentIdConfig
     {
+        private readonly Lazy<Dictionary<string, string>> _reversedDepotMappings;
+        private readonly Lazy<Dictionary<string, string>> _reversedDirectoryMappings;
```
Comment (yufeih) [inline src/docfx/config/DocumentIdConfig.cs]: nit: put these computed values to `Docset` to align with other computed properties like _Redirections_, _RestoreMap_

File: src/docfx/config/DocumentIdConfig.cs [hunk 3]
```diff
@@ -14,5 +21,71 @@ internal class DocumentIdConfig
...
+                : normalizedFilePathToSourceBasePath;
+
+            return (depotName, mappedPathRelativeToSourceBasePath);
+        }
+
+        private static (string value, string matchedDirectory) GetReversedMapping(Dictionary<string, string> mappings, string normalizedFilePathToSourceBasePath)
+        {
+            foreach (var (path, value) in mappings)
+            {
+                if (string.Equals(path, normalizedFilePathToSourceBasePath, PathUtility.PathComparison))
```
Comment (yufeih) [inline src/docfx/config/DocumentIdConfig.cs]: we probably need a `PathUtility.PathMatches` method to detect if a path is inside a folder or is the same path. `routes` also need this.

Comment (yufeih) [reply src/docfx/config/DocumentIdConfig.cs]: Also, it is proved problematic to indicate folders using trailing slash, https://github.com/dotnet/docfx/issues/3077.

File: docs/specs/metadata.yml (hunk 1 above)
Comment (OsmondJiang) [reply docs/specs/metadata.yml]: the original syntax is hard to understand and a little confused. and also it contains a lot of duplications. I prefer to splite them and reverse them

---
In reply to: [203581517](https://github.com/dotnet/docfx/pull/3095#discussion_r203581517) [](ancestors = 203581517)

File: src/docfx/config/DocumentIdConfig.cs (hunk 2 above)
Comment (OsmondJiang) [reply src/docfx/config/DocumentIdConfig.cs]: I followed the RouteConfig which only want to provide two methods for document id config.

---
In reply to: [203581997](https://github.com/dotnet/docfx/pull/3095#discussion_r203581997) [](ancestors = 203581997)

File: src/docfx/config/DocumentIdConfig.cs (hunk 3 above)
Comment (OsmondJiang) [reply src/docfx/config/DocumentIdConfig.cs]: OK, i will add one.

---
In reply to: [203582744](https://github.com/dotnet/docfx/pull/3095#discussion_r203582744) [](ancestors = 203582744)

Comment (OsmondJiang) [reply src/docfx/config/DocumentIdConfig.cs]: Let's resolve this issue together, for this PR, i follow the original rules, folder need end with '/'

---
In reply to: [203583022](https://github.com/dotnet/docfx/pull/3095#discussion_r203583022) [](ancestors = 203583022)

File: src/docfx/config/RouteConfig.cs [hunk 4]
```diff
@@ -28,20 +27,19 @@ public string GetOutputPath(string path)
         {
             Debug.Assert(!string.IsNullOrEmpty(path));
 
-            if (Source == "./")
+            if (PathUtility.Match(path, Source))
             {
-                return Path.Combine(Destination, path);
-            }
+                if (Source == "./")
```
Comment (yufeih) [inline src/docfx/config/RouteConfig.cs]: put "./" check under `PathUtility.Match`

add UT

`PathUtility.Match` can return `(bool matches, bool isFile, string remainingPath)` like:

- `Match("a", "a") -> (true, true, "")`
- `Match("a/b", "a/") -> (true, false, "b")`

**Pull request merged**